src/compiler/scala/tools/nsc/Reporting.scala

```diff
@@ -14,102 +14,212 @@ package scala
+        }
+      }
+    }
 
-    // behold! the symbol that caused the deprecation warning (may not be deprecated itself)
-    def deprecationWarning(pos: Position, sym: Symbol, msg: String, since: String): Unit = _deprecationWarnings.warn(pos, msg, since)
-    def deprecationWarning(pos: Position, sym: Symbol): Unit = {
-      val version = sym.deprecationVersion.getOrElse("")
+    private def siteName(s: Symbol) = if (s.exists) s.fullNameString else ""
+
+    def deprecationWarning(pos: Position, msg: String, since: String, site: String, origin: String): Unit =
+      checkSuppressedAndIssue(Message.Deprecation(pos, msg, site, origin, Version.fromString(since)))
+
+    def deprecationWarning(pos: Position, origin: Symbol, site: Symbol, msg: String, since: String): Unit =
+      deprecationWarning(pos, msg, since, siteName(site), siteName(origin))
+
+    def deprecationWarning(pos: Position, origin: Symbol, site: Symbol): Unit = {
+      val version = origin.deprecationVersion.getOrElse("")
       val since   = if (version.isEmpty) version else s" (since $version)"
-      val message = sym.deprecationMessage.map(": " + _).getOrElse("")
-      deprecationWarning(pos, sym, s"$sym${sym.locationString} is deprecated$since$message", version)
+      val message = origin.deprecationMessage.map(": " + _).getOrElse("")
+      deprecationWarning(pos, origin, site, s"$origin${origin.locationString} is deprecated$since$message", version)
     }
 
     private[this] var reportedFeature = Set[Symbol]()
-    def featureWarning(pos: Position, featureName: String, featureDesc: String, featureTrait: Symbol, construct: => String = "", required: Boolean): Unit = {
+    // we don't have access to runDefinitions here, so mapping from strings instead of feature symbols
+    private val featureCategory: Map[String, WarningCategory.Feature] = {
+      import WarningCategory._
+      Map(
+        ("dynamics", FeatureDynamics),
+        ("existentials", FeatureExistentials),
+        ("higherKinds", FeatureHigherKinds),
+        ("implicitConversions", FeatureImplicitConversions),
+        ("postfixOps", FeaturePostfixOps),
+        ("reflectiveCalls", FeatureReflectiveCalls),
+        ("macros", FeatureMacros)
+      ).withDefaultValue(Feature)
+    }
+    def featureWarning(pos: Position, featureName: String, featureDesc: String, featureTrait: Symbol, construct: => String = "", required: Boolean, site: Symbol): Unit = {
       val req     = if (required) "needs to" else "should"
       val fqname  = "scala.language." + featureName
       val explain = (
         if (reportedFeature contains featureTrait) "" else
-        s"""|
-            |----
-            |This can be achieved by adding the import clause 'import $fqname'
-            |or by setting the compiler option -language:$featureName.
-            |See the Scaladoc for value $fqname for a discussion
-            |why the feature $req be explicitly enabled.""".stripMargin
-      )
+          s"""
```

Comment:
Maybe it's easier to achieve the desired indentation alignment with `sm"""` because of the extra letter. Is it appealing with vbar in even column, or to left of even column?

Reply:
I don't have a personal opinion here, my only opinion is "do what ever IntelliJ does", so I don't get any spurious changes when editing this code (in IntelliJ)...